docs/src/main/tut/json.md

```diff
@@ -0,0 +1,182 @@
+
+```tut:book
+import argonaut._, Argonaut._, ArgonautShapeless._
+import org.http4s.argonaut._
+
+import org.http4s._, org.http4s.dsl._
+
+case class User(name: String)
+case class Hello(greeting: String)
+
+val jsonService = HttpService {
+  case r @ POST -> Root / "hello" =>
+    r.decode[User](user =>
+      Ok(Hello(s"Hello, ${user.name}"))(jsonEncoderOf)
+    )(jsonOf)
+}
+
+import org.http4s.server.syntax._
+import org.http4s.server.blaze._
+val builder = BlazeBuilder.bindHttp(8080, "localhost").mountService(jsonService, "/")
+val server = builder.run
+```
+
+## Receiving JSON
+
+The `EntityDecoder[T]` instances work similarly to the `EntityEncoder[T]`
+instances, except it's `jsonOf` to create the instances.
+
+Let's talk to the server defined above.
+
+```tut:book
+import org.http4s.client.blaze._
+import org.http4s.Uri
+
+val httpClient = PooledHttp1Client()
+val req = Request(uri = Uri.uri("http://localhost:8080/hello"), method = Method.POST).withBody(User("Anabelle"))(jsonEncoderOf)
+httpClient.expect(req)(jsonOf[Hello]).run
+```
+
+And clean everything up.
+
+```tut:book
+httpClient.shutdownNow()
+server.shutdownNow()
+```
+
+## Talking to the Github API
+
+As an example, we'll use the [github-orgs] endpoint. The data is shortened a bit
+for clarity.
+
+```json
+[
+  {
+    "id": 3692188,
+    "name": "http4s",
+    "full_name": "http4s/http4s",
+    "owner": {
+      "login": "http4s",
+      "id": 1527492
+      // ...
+    },
+    "private": false,
+    "html_url": "https://github.com/http4s/http4s",
+    "description": "A minimal, idiomatic Scala interface for HTTP",
+    "size": 21365,
+    "stargazers_count": 533,
+    "language": "Scala",
+    "forks_count": 118,
+    "open_issues_count": 69,
+    "forks": 118,
+    "open_issues": 69,
+    "watchers": 533
+    // ...
+  }
+  // ...
+]
+```
+
+To capture this structure, we'll need corresponding case classes.
+
+```tut:book
+case class User(login: String, id: Long)
+case class Repo(id: Long, name: String, full_name: String, owner: User, `private`: Boolean, html_url: String, description: String, size: Int, stargazers_count: Int, language: String, forks_count: Int, open_issues_count: Int, forks: Int, open_issues: Int, watchers: Int)
+```
+
+This parts skips over the [client] explanation. We'll use argonaut, with
+[argonaut-shapeless] for codec derivation. The json decoder is provided by
+`ArgonautShapeless`, and passed to the `client.expect` method via the second
+argument, which is usually an implicit `EntityDecoder`, but in this case, we
+want to be more explicit.
+
+<!-- For more information about the uri templating, visit [uri]. -->
+
+```tut:book
+import scalaz.concurrent.Task
+
+import org.http4s.util.CaseInsensitiveString.ToCaseInsensitiveStringSyntax
+import org.http4s.UriTemplate
+import org.http4s.UriTemplate._
+
+val httpClient = PooledHttp1Client()
+
+def repos(org: String): Task[List[Repo]] = {
+  val uri = UriTemplate(
```

Comment:
I had completely forgotten we have `UriTemplate` support.  It would be a lot more exciting if we could parse from Strings.

`Uri.fromString(s"https://api.github.com/orgs/${org}/repos").valueOr(throw _)` is a lot more readable to me.  That `valueOr(throw _)` is something we need to make more convenient to deal with.  Our rigidity in error handling URIs borders on the impractical.